c10/util/TypeCast.h

```diff
@@ -68,6 +69,36 @@ struct static_cast_with_inter_type<uint8_t, src_t> {
   }
 };
 
+template <>
+struct static_cast_with_inter_type<c10::complex<c10::Half>, c10::BFloat16> {
+  C10_HOST_DEVICE __ubsan_ignore_undefined__ static inline c10::complex<
+      c10::Half>
+  apply(c10::BFloat16 src) {
+    return static_cast<c10::complex<c10::Half>>(c10::complex<float>{src});
+  }
+};
+
+template <>
+struct static_cast_with_inter_type<c10::complex<c10::Half>, c10::Half> {
+  C10_HOST_DEVICE __ubsan_ignore_undefined__ static inline c10::complex<
+      c10::Half>
+  apply(c10::Half src) {
+    return static_cast<c10::complex<c10::Half>>(c10::complex<float>{src});
+  }
+};
+
+template <>
+struct static_cast_with_inter_type<
+    c10::complex<c10::Half>,
+    c10::complex<double>> {
+  C10_HOST_DEVICE __ubsan_ignore_undefined__ static inline c10::complex<
+      c10::Half>
+  apply(c10::complex<double> src) {
+    return static_cast<c10::complex<c10::Half>>(
```

Comment:
why can't we directly cast complex half to complex double?

Reply:
Currently, we don't have a constructor for `complex<Half>` which takes `complex<double>`. Having that constructor leads to ambiguos resolution while converting from `double` to `complex<Half>` reason being both constructors which take `complex<float>` and `complex<double>` will be equally valid (as both `complex<float>` and `complex<double>` can be constructed from a double).

```
../c10/util/TypeCast.h: In instantiation of 'static dest_t c10::static_cast_with_inter_type<dest_t, src_t>::apply(src_t) [with dest_t = c10::complex<c10::Half>; src_t = double]':
../c10/util/TypeCast.h:212:54:   required from 'To c10::convert(From) [with To = c10::complex<c10::Half>; From = double]'
../c10/util/TypeCast.h:224:27:   required from 'To c10::checked_convert(From, const char*) [with To = c10::complex<c10::Half>; From = double]'
../c10/core/Scalar.h:70:3:   required from here
../c10/util/TypeCast.h:49:12: error: call of overloaded 'complex(double&)' is ambiguous
   49 |     return static_cast<dest_t>(r);
      |            ^~~~~~~~~~~~~~~~~~~~~~
In file included from ../c10/core/ScalarType.h:5,
                 from ../c10/core/Scalar.h:11,
                 from aten/src/ATen/core/TensorBody.h:16,
                 from ../aten/src/ATen/core/jit_type.h:5,
                 from ../aten/src/ATen/core/function_schema.h:6,
                 from ../aten/src/ATen/core/op_registration/infer_schema.h:8,
                 from ../torch/library.h:61,
                 from ../aten/src/ATen/BatchingRegistrations.cpp:1:
../c10/util/Half.h:409:35: note: candidate: 'c10::complex<c10::Half>::complex(const c10::complex<double>&)'
  409 |   C10_HOST_DEVICE explicit inline complex(const c10::complex<double>& value)
      |                                   ^~~~~~~
../c10/util/Half.h:407:35: note: candidate: 'c10::complex<c10::Half>::complex(const c10::complex<float>&)'
  407 |   C10_HOST_DEVICE explicit inline complex(const c10::complex<float>& value)
      |                                   ^~~~~~~
```

I am also planning to look at all the constructors which `complex<Half>` should have. Currently it also lacks a constructor from `Half` to `complex<Half>` i.e. from real to complex.

Reply:
oh cool thanks for clarifying!

> I am also planning to look at all the constructors which complex<Half> should have. Currently it also lacks a constructor from Half to complex<Half> i.e. from real to complex.

good idea! a good thing to add in the tracker as well